#!/usr/bin/env python
PACKAGE = "mav_ilqr_control"

from dynamic_reconfigure.parameter_generator_catkin import *

gen = ParameterGenerator()

gen.add("phi_gain",    double_t,    0, "proportional gain of roll angle",   1.0 ,  0.1, 10000.0)
gen.add("theta_gain",   double_t,    0, "proportional gain of pitch angle",  20.0 ,  0.1, 10000.0)
gen.add("psi_gain",   double_t,    0, "proportional gain of yaw angle",  1.0 ,  0.1, 10000.0)

gen.add("x_gain",    double_t,    0, "proportional gain of x axis",   3 ,  0.1, 10000.0)
gen.add("vx_gain",   double_t,    0, "derivative gain of x axis",  0.1 ,  0.1, 10000.0)
gen.add("y_gain",   double_t,    0, "proportional gain of y axis",  3 ,  0.1, 10000.0)
gen.add("vy_gain",    double_t,    0, "derivative gain of y axis",   0.1 ,  0.1, 10000.0)
gen.add("z_gain",   double_t,    0, "proportional gain of z axis",  15 ,  0.1, 10000.0)
gen.add("dz_gain",   double_t,    0, "derivative gain of z axis",  0.3 ,  0.1, 10000.0)

gen.add("thrust_gain",   double_t,    0, "thrust control weight",  0.1 ,  0.1, 1000.0)
gen.add("roll_ref_gain",   double_t,    0, "roll control weight",  5.0 ,  0.1, 1000.0)
gen.add("pitch_ref_gain",    double_t,    0, "pitch control weight",   10.0 ,  0.1, 1000.0)
gen.add("yaw_rate_command_gain",   double_t,    0, "yaw_rate control weight",  10.0 ,  0.1, 1000.0)
gen.add("alpha_gain",   double_t,    0, "tilt control weight",  5 ,  0.1, 1000.0)

gen.add("K_yaw",   double_t,    0, "yaw gain", 10 ,  0.01, 1000.0)
gen.add("yaw_rate_limit",   double_t,    0, "yaw rate command limit", 10 ,  0.01, 10.0)

gen.add("speed_limit_on",   bool_t,   0, "active speed safety limit",  False)
gen.add("vx_limit",   double_t,    0, "vx limit", 5 ,  0, 1000.0)
gen.add("vy_limit",   double_t,    0, "vy limit", 5 ,  0, 1000.0)
gen.add("vz_limit",   double_t,    0, "vz limit", 5 ,  0, 1000.0)

gen.add("position_limit_on",   bool_t,   0, "active position safety limit",  True)
gen.add("x_limit_inf",   double_t,    0, "x position lower limit", -1.58 ,  -10, 5)
gen.add("x_limit_sup",   double_t,    0, "x position upper limit", 1.48 ,  -5, 10)
gen.add("y_limit_inf",   double_t,    0, "y position lower limit", -1 ,  -10, 5)
gen.add("y_limit_sup",   double_t,    0, "y position upper limit", 1.7 ,  -5, 10)
gen.add("z_limit_inf",   double_t,    0, "z position lower limit", 0 ,  -10, 5)
gen.add("z_limit_sup",   double_t,    0, "z position upper limit", 1.77 ,  -5, 10)


exit(gen.generate(PACKAGE, "Config", "iLQRController"))
